### ISSUE TITLE
Operator doesn't use new Pulumi config

### ISSUE THREAD

**Issue by skdltmxn** [pulumi/pulumi-kubernetes-operator#411]
### What happened?

Since I updated my Pulumi Kubernetes Operator to 1.11.1, it seems like the operator cannot fetch newly updated Pulumi config from git repo.

Say, I have following config:

```yaml
app:
  name: my-service
  repo: my-secret-repo
  version: v1
```

and, if I update version to `v2`, the operator cannot recognize that the version has been modified.
It keeps applying with `v1`.  
Only after I kill the operator pod and then newly spawned pod starts to see `v2`.

This doesn't happen in v1.11.0.

### Expected Behavior

The operator should correctly apply newly updated Pulumi config

### Steps to reproduce

1. Deploy Pulumi Kubernetes Operator v1.11.1
2. Sync Kubernetes using the initial Pulumi config
3. Update Pulumi config in any way.
4. The operator doesn't see new config

### Output of `pulumi about`

```
CLI
Version      3.46.0
Go Version   go1.19.2
Go Compiler  gc

Plugins
NAME        VERSION
aws         5.29.1
azuread     5.34.0
kubernetes  3.24.0
nodejs      unknown

Host
OS       debian
Version  11.5
Arch     x86_64

This project is written in nodejs: executable='/usr/bin/node' version='v16.18.0'

Pulumi locates its logs in /tmp by default
warning: Failed to get information about the Pulumi program's dependencies: could not find either /tmp/pulumi-working/pulumi/pulumi-central-stack/yarn.lock or /tmp/pulumi-working/pulumi/pulumi-central-stack/package-lock.json
warning: A new version of Pulumi is available. To upgrade from version '3.46.0' to '3.54.0', visit https://pulumi.com/docs/reference/install/ for manual instructions and release notes.
```

I ran the command directly from the pod

### Additional context

This is my Stack with some irrelevant values redacted.

```yaml
apiVersion: pulumi.com/v1
kind: Stack
metadata:
  labels:
    app.kubernetes.io/instance: pulumi-central-stack
    app.kubernetes.io/managed-by: pulumi
    app.kubernetes.io/name: pulumi-central-stack
  name: pulumi-central-stack
  namespace: pulumi
spec:
  backend: s3://XXX
  branch: refs/heads/main
  continueResyncOnCommitMatch: true
  destroyOnFinalize: false
  gitAuth:
    accessToken:
      secret:
        key: github_token
        name: pulumi-github-token
      type: Secret
  projectRepo: https://XXXX
  refresh: true
  resyncFrequencySeconds: 60
  stack: central
```

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by skdltmxn** [comment]
Yes it is.
I see following logs from the operator pod.

```
{"level":"debug","ts":"2023-02-13T14:31:48.171Z","logger":"controller_stack","msg":"Initial autostack config","Request.Namespace":"pulumi","Request.Name":"pulumi-dev-stack","config": "Config with v2 is printed"}
{"level":"debug","ts":"2023-02-13T14:31:48.174Z","logger":"controller_stack","msg":"stackConfig loaded","Request.Namespace":"pulumi","Request.Name":"pulumi-dev-stack","stack":{},"stackConfig": "Config with v1 is printed"} 
```

I ran `git diff` command in the operator pod while it's running, and it showed as if the config file has been reverted back to v1.

**Comment by squaremo** [comment]
I can reproduce this. I did these steps:

 - use a project which was created with `pulumi new lubernetes-typescript`, and adapt it to use a config item `tag` (in the image used in the deployment)
 - do `pulumi login file://$PWD` to use a file backend (this makes it easier to name the stack and config)
 - do `pulumi stack` and create a new stack `dev`
 - do `pulumi config set tag "1.22"` to set the config in `Pulumi.dev.yaml`
 - commit `Pulumi.dev.yaml` to the project git repo
 - create a stack object that points at the git repo, and uses a file backend, and the stack name `dev`
 - observe that it deploys using the tag value of `1.22`
 - change the tag value to `1.23` in Pulumi.dev.yaml` commit the change, and push
 - wait until the operator processes the stack again, and observe that it shows the new value `1.23` in one log line, then the old value `1.22` in another, and reports no change

Now I'm trying to find when this behaviour changed.

**Comment by squaremo** [comment]
OK. So yes, it is due to the change in #407, in an unexpected way. The project and stack file loader in pulumi/pulumi caches results by path:
https://github.com/pulumi/pulumi/blob/master/sdk/go/common/workspace/loaders.go#L79. This change was introduced in https://github.com/pulumi/pulumi/issues/6564 in response to performance concerns, it looks like.

However, in a long-running process like the operator which expects to run stacks repeatedly, this causes problems. While the stack was checked out to fresh directory each time (before #407), it was just a memory leak (perhaps causing or at least contributing to #364). But now stacks are built at stable paths, it means changes made in a Pulumi.yaml or Pulumi.<stack>.yaml file in the git repo (or Flux source) will not be respected.

I'm not sure what to do about this, short of changes upstream to pulumi/pulumi; stable paths is really the best, and probably only, solution to #381.

**Comment by squaremo** [comment]
As a workaround, you could put the config you want to change into the Stack object, which will get loaded every time. This has a limitation though: it's just key/value pairs, so not structured config, and you might have to alter your program(s) to suit.

In the meantime, we're looking into #12152 above, and I'll see if there's anything that can be done here.

**Comment by skdltmxn** [comment]
Actually I have reverted my operator back to 1.11.0 and am willing to update it when #12152 is resolved

**Comment by squaremo** [comment]
The caching has been removed in pulumi SDK v3.57.0: https://github.com/pulumi/pulumi/releases/tag/v3.57.0 (the specific pull request was https://github.com/pulumi/pulumi/pull/12370). I fully expect that to fix our problem here -- I'll bump the version used here, and see.

**Comment by squaremo** [comment]
Released in v1.11.2. I checked locally that my repro (https://github.com/pulumi/pulumi-kubernetes-operator/issues/411#issuecomment-1428114730) no longer repros. If you try the new release, let me know if you still get this problem, and if you see the memory leak in #420. v1.11.1 (which had _this_ problem) fixed at least one memory leak.
